multinet/multinet/db.py

```diff
@@ -1,11 +1,30 @@
 """Low-level database operations to fulfill GraphQL queries."""
 import os
+import sys
 
 from arango import ArangoClient
+from requests.exceptions import ConnectionError
+
+from flask import current_app as app
 
 from multinet.types import Row, Entity, EntityType, Cursor
 
 
+def test_db(arango):
```

Comment:
Let's use a more descriptive function name, something like `db_required`.

Then this function could be used as a decorator for functions that require the database, and it could automatically return the right status code if the database isn't running.

Or, perhaps more likely, this function could be called within `with_client()`, since that decorator already implies the idea of "database required".

Reply:
I changed the name to the one you suggested. Are you imagining that the db_required function will live inside the with_client() function and then we'd call it in the wrapper function? If we're good, call fun(), otherwise just return?

Reply:
I think so, but rather than "just return", the cool thing would be to raise the 500 error at that point. I think that could just be done with `abort()` since there's no data payload in that case, but I'm also not entirely certain that will work, since we're not obviously within a flask context there.

Reply:
Aborting there is giving the same graphql error where it prints the whole stack trace and is un-catchable. I think we should `return None` here and then in multinet.py if the data is None, we can `abort(500)`. That way if the db is up, we don't get a 500 and if the db is down we do. What do you think?

Reply:
I want to avoid doing this so we don't have to check for `None` in a bunch of places throughout the API. Let me think about this a bit.

Reply:
I want to avoid doing this so we don't have to check for None in a bunch of places throughout the API. Let me think about this a bit.

Reply:
Ok, so I think I've figured this out.

We don't want to check if the DB is live on every single DB operation (which is what we'd be doing if we rolled this check into `with_client`); instead we just want to check once on any endpoint that will engage the DB. So I think `db_required` should become a decorator that performs the DB liveness test, does `return (500, "Database Not Live")` if the test fails, and runs the wrapped function if the test passes. Since the context is a Flask route, we should be able to do this. And since that liveness testing operation requires the use of the database, it can be wrapped in the `with_client` decorator.

So in the end, our mistake was in thinking that `with_client` should always include `db_required`, whereas it's actually the other way around.

Let me know what you think of this.

Reply:
Edited the blueprints to run the require_db function before each route.